File: qdrant_interpreter_plugin/index_manager.py
"""
Qdrant Index Manager
====================

High-level collection management built on top of :class:`UseCaseParamAdvisor`.
The manager wraps a :class:`qdrant_client.QdrantClient` and provides
auto-parameterised CRUD, payload indexing, health checks, and optimisation
in a single, easy-to-use API.

Typical agent usage::

    from qdrant_client import QdrantClient
    from qdrant_interpreter_plugin.index_manager import QdrantIndexManager

    manager = QdrantIndexManager(QdrantClient(":memory:"))

    # Create with auto-selected parameters
    result = manager.auto_create(
        collection_name="products",
        use_case="semantic search over product descriptions using OpenAI ada-002",
        expected_count=500_000,
        priority="recall",
    )
    # result["distance"] == "Cosine", result["quantization"] == "scalar", ...

    # Add a payload index for fast filtered queries
    manager.create_payload_index("products", "category", "keyword")

    # Inspect health and get recommendations
    health = manager.health("products")
"""

from __future__ import annotations

from typing import Any

from qdrant_client import QdrantClient
from qdrant_client.models import (
    PayloadSchemaType,
    SparseIndexParams,
    SparseVectorParams,
    VectorParams,
)

from .param_advisor import PriorityHint, UseCaseParamAdvisor


class QdrantIndexManager:
    """Manages Qdrant collections with automatic parameter selection.

    Wraps a :class:`qdrant_client.QdrantClient` and adds a layer of
    use-case-aware intelligence via :class:`UseCaseParamAdvisor`. All
    operations return plain dicts so they integrate cleanly with LangChain
    tools and agent loops.

    Parameters
    ----------
    client:
        An already-constructed ``QdrantClient`` instance. The manager does
        not take ownership of the client's lifetime.
    """

    def __init__(self, client: QdrantClient) -> None:
        self.client = client
        self._advisor = UseCaseParamAdvisor()

    # ------------------------------------------------------------------
    # Collection lifecycle
    # ------------------------------------------------------------------

    def auto_create(
        self,
        collection_name: str,
        use_case: str,
        vector_size: int | None = None,
        expected_count: int = 100_000,
        priority: PriorityHint = "recall",
        memory_gb_available: float | None = None,
        sparse_vectors: bool = False,
        overrides: dict[str, Any] | None = None,
    ) -> dict[str, Any]:
        """Create a collection using automatically recommended parameters.

        Parameters
        ----------
        collection_name:
            Name for the new collection.
        use_case:
            Natural-language description of the intended use, e.g.
            ``"semantic search over product descriptions using OpenAI ada-002"``.
        vector_size:
            Vector dimension. Inferred from ``use_case`` keywords if not given.
        expected_count:
            Approximate number of vectors; influences HNSW and quantization.
        priority:
            Optimisation goal: ``"recall"``, ``"latency"``, or ``"memory"``.
        memory_gb_available:
            Available RAM in GB; drives quantization aggressiveness.
        sparse_vectors:
            When ``True``, also configure a ``"sparse"`` named sparse-vector
            field for hybrid (dense + sparse) search.
        overrides:
            Dict of parameter overrides merged on top of the advisor
            recommendations. Recognised keys:

            * ``"vector_size"`` — override inferred dimension
            * ``"distance"`` — a :class:`qdrant_client.models.Distance` value
            * ``"on_disk_vectors"`` — bool, store vectors on disk
            * ``"on_disk_payload"`` — bool
            * ``"quantization_config"`` — a quantization config object

        Returns
        -------
        dict
            Summary of the applied parameters and the advisor's rationale.
            Keys: ``collection_name``, ``vector_size``, ``distance``,
            ``quantization``, ``hnsw_m``, ``hnsw_ef_construct``,
            ``on_disk_payload``, ``sparse_vectors``, ``rationale``.
        """
        advice = self._advisor.advise(
            use_case=use_case,
            vector_size=vector_size,
            expected_count=expected_count,
            priority=priority,
            memory_gb_available=memory_gb_available,
        )
        overrides = overrides or {}

        # Sparse vector configuration
        sparse_cfg: dict[str, SparseVectorParams] | None = None
        if sparse_vectors:
            sparse_cfg = {
                "sparse": SparseVectorParams(
                    index=SparseIndexParams(on_disk=False)
                )
            }

        self.client.create_collection(
            collection_name=collection_name,
            vectors_config=VectorParams(
                size=overrides.get("vector_size", advice.vector_size),
                distance=overrides.get("distance", advice.distance),
                on_disk=overrides.get("on_disk_vectors", False),
            ),
            sparse_vectors_config=sparse_cfg,
            hnsw_config=advice.hnsw_config,
            optimizers_config=advice.optimizers_config,
            quantization_config=overrides.get(
                "quantization_config", advice.quantization_config
            ),
            on_disk_payload=overrides.get("on_disk_payload", advice.on_disk_payload),
        )

        return {
            "collection_name": collection_name,
            "vector_size": advice.vector_size,
            "distance": advice.distance.value,
            "quantization": advice.quantization_mode,
            "hnsw_m": advice.hnsw_config.m,
            "hnsw_ef_construct": advice.hnsw_config.ef_construct,
            "on_disk_payload": advice.on_disk_payload,
            "sparse_vectors": sparse_vectors,
            "rationale": advice.rationale,
        }

    def ensure(
        self,
        collection_name: str,
        use_case: str,
        vector_size: int | None = None,
        **kwargs: Any,
    ) -> dict[str, Any]:
        """Create the collection only if it does not already exist.

        Returns a dict with an ``"action"`` key set to either ``"created"``
        or ``"already_exists"``.
        """
        existing = {c.name for c in self.client.get_collections().collections}
        if collection_name in existing:
            info = self.client.get_collection(collection_name)
            return {
                "action": "already_exists",
                "collection_name": collection_name,
                "points_count": info.points_count or 0,
                "status": info.status.value,
            }
        result = self.auto_create(collection_name, use_case, vector_size, **kwargs)
        result["action"] = "created"
        return result

    def delete(self, collection_name: str) -> bool:
        """Delete a collection. Returns ``True`` on success."""
        self.client.delete_collection(collection_name)
        return True

    # ------------------------------------------------------------------
    # Payload indexes
    # ------------------------------------------------------------------

    _SCHEMA_TYPE_MAP: dict[str, PayloadSchemaType] = {
        "keyword": PayloadSchemaType.KEYWORD,
        "integer": PayloadSchemaType.INTEGER,
        "float": PayloadSchemaType.FLOAT,
        "geo": PayloadSchemaType.GEO,
        "text": PayloadSchemaType.TEXT,
        "bool": PayloadSchemaType.BOOL,
        "datetime": PayloadSchemaType.DATETIME,
    }

    def create_payload_index(
        self,
        collection_name: str,
        field_name: str,
        field_type: str = "keyword",
    ) -> dict[str, str]:
        """Create a payload index for fast filtered search.

        Indexed payload fields enable WHERE-style filtering with near-zero
        extra latency, even on collections with millions of vectors.

        Parameters
        ----------
        collection_name:
            Target collection.
        field_name:
            Payload key to index (e.g. ``"category"``, ``"user_id"``).
        field_type:
            Schema type: ``keyword`` | ``integer`` | ``float`` | ``geo`` |
            ``text`` | ``bool`` | ``datetime``.

        Returns
        -------
        dict
            ``{"collection", "field", "type", "status"}``.

        Raises
        ------
        ValueError
            When ``field_type`` is not a recognised schema type.
        """
        schema_type = self._SCHEMA_TYPE_MAP.get(field_type.lower())
        if schema_type is None:
            valid = list(self._SCHEMA_TYPE_MAP)
            raise ValueError(
                f"Unknown field_type '{field_type}'. Valid options: {valid}"
            )
        self.client.create_payload_index(
            collection_name=collection_name,
            field_name=field_name,
            field_schema=schema_type,
        )
        return {
            "collection": collection_name,
            "field": field_name,
            "type": field_type,
            "status": "indexed",
        }

    # ------------------------------------------------------------------
    # Health and optimisation
    # ------------------------------------------------------------------

    def health(self, collection_name: str) -> dict[str, Any]:
        """Return runtime stats and parameter recommendations for a collection.

        Returns
        -------
        dict
            Keys: ``collection_name``, ``points_count``,
            ``indexed_vectors_count``, ``status``, ``optimizer_status``,
            ``vector_size`` (if detectable), ``distance`` (if detectable),
            ``recommendation`` (optional advisory message).
        """
        info = self.client.get_collection(collection_name)

        result: dict[str, Any] = {
            "collection_name": collection_name,
            "points_count": info.points_count or 0,
            "indexed_vectors_count": info.indexed_vectors_count or 0,
            "status": info.status.value,
            "optimizer_status": str(info.optimizer_status),
        }

        # Decode vector config — can be VectorParams or dict[str, VectorParams]
        cfg = info.config.params.vectors
        if isinstance(cfg, VectorParams):
            result["vector_size"] = cfg.size
            result["distance"] = cfg.distance.value
        elif isinstance(cfg, dict) and cfg:
            first = next(iter(cfg.values()))
            result["vector_size"] = first.size
            result["distance"] = first.distance.value

        # Advisory message for collections that would benefit from quantization
        quant = info.config.quantization_config
        count = info.points_count or 0
        if quant is None and count > 100_000:
            result["recommendation"] = (
                f"Collection has {count:,} vectors but no quantization configured. "
                "Consider scalar INT8 quantization for ~4× memory reduction with "
                "minimal recall impact."
            )

        return result

    def optimize(
        self, collection_name: str, priority: PriorityHint = "recall"
    ) -> dict[str, Any]:
        """Apply optimised indexing/memmap thresholds to an existing collection.

        Does not modify vector data or quantization settings. Safe to call
        on a live collection.

        Returns
        -------
        dict
            Applied optimizer settings.
        """
        count = (
            self.client.get_collection(collection_name).points_count or 100_000
        )
        optimizers = self._advisor._pick_optimizers(count, priority)
        self.client.update_collection(
            collection_name=collection_name,
            optimizer_config=optimizers,
        )
        return {
            "collection": collection_name,
            "priority": priority,
            "applied_optimizers": {
                "indexing_threshold": optimizers.indexing_threshold,
                "memmap_threshold": optimizers.memmap_threshold,
            },
        }

File: ts-plugin/src/indexManager.ts
/**
 * QdrantManager
 * =============
 * High-level Qdrant collection management powered by UseCaseParamAdvisor.
 * All methods return plain objects so they compose cleanly with LangChain
 * tools and interpreter code.
 */

import { QdrantClient } from "@qdrant/js-client-rest";
import { UseCaseParamAdvisor, pickOptimizers } from "./paramAdvisor.js";
import type { PriorityHint, CollectionParamAdvice } from "./paramAdvisor.js";

export type { PriorityHint };

// ---------------------------------------------------------------------------
// Public option types
// ---------------------------------------------------------------------------

export interface AutoCreateOptions {
  /** Free-text description of the workload — drives parameter selection. */
  useCase: string;
  /** Embedding dimension. Inferred from useCase keywords when omitted. */
  vectorSize?: number;
  /** Approximate vector count (default 100 000). */
  expectedCount?: number;
  /** Optimisation goal (default "recall"). */
  priority?: PriorityHint;
  /** Available RAM in GB; triggers aggressive quantization below 4 GB. */
  memoryGbAvailable?: number;
  /** Add a sparse vector slot for hybrid (dense + sparse) search. */
  sparseVectors?: boolean;
}

export interface CollectionSummary {
  collection_name: string;
  vector_size: number;
  distance: string;
  quantization: string;
  hnsw_m: number | undefined;
  hnsw_ef_construct: number | undefined;
  on_disk_payload: boolean;
  sparse_vectors: boolean;
  rationale: Record<string, string>;
  action?: string;
  points_count?: number;
  status?: string;
}

export interface HealthInfo {
  collection_name: string;
  points_count: number;
  indexed_vectors_count: number;
  status: string;
  optimizer_status: string;
  vector_size?: number;
  distance?: string;
  recommendation?: string;
}

export interface PayloadIndexResult {
  collection: string;
  field: string;
  type: string;
  status: string;
}

export interface OptimizeResult {
  collection: string;
  priority: string;
  applied_optimizers: { indexing_threshold: number; memmap_threshold: number };
}

// ---------------------------------------------------------------------------
// Field-type map (validated at call-site)
// ---------------------------------------------------------------------------

const FIELD_TYPES: Record<string, true> = {
  keyword: true,
  integer: true,
  float: true,
  geo: true,
  text: true,
  bool: true,
  datetime: true,
};

const VALID_FIELD_TYPES = Object.keys(FIELD_TYPES).join(", ");

type PayloadFieldType = "keyword" | "integer" | "float" | "geo" | "text" | "bool" | "datetime";

function assertFieldType(raw: string): PayloadFieldType {
  if (!FIELD_TYPES[raw]) {
    throw new Error(`Unknown field_type '${raw}'. Valid options: ${VALID_FIELD_TYPES}`);
  }
  return raw as PayloadFieldType;
}

// ---------------------------------------------------------------------------
// QdrantManager
// ---------------------------------------------------------------------------

export class QdrantManager {
  readonly client: QdrantClient;
  private readonly advisor = new UseCaseParamAdvisor();

  constructor(client: QdrantClient) {
    this.client = client;
  }

  // ------------------------------------------------------------------ //
  // Collection lifecycle                                                //
  // ------------------------------------------------------------------ //

  async autoCreate(name: string, options: AutoCreateOptions): Promise<CollectionSummary> {
    const {
      useCase,
      vectorSize,
      expectedCount = 100_000,
      priority = "recall",
      memoryGbAvailable,
      sparseVectors = false,
    } = options;

    const advice = this.advisor.advise({
      useCase,
      vectorSize,
      expectedCount,
      priority,
      memoryGbAvailable,
    });

    await this.client.createCollection(name, {
      vectors: {
        size: advice.vector_size,
        distance: advice.distance,
        on_disk: false,
      },
      hnsw_config: advice.hnsw_config,
      optimizers_config: advice.optimizers_config,
      on_disk_payload: advice.on_disk_payload,
      ...(advice.quantization_config !== null
        ? { quantization_config: advice.quantization_config }
        : {}),
      ...(sparseVectors
        ? { sparse_vectors: { sparse: { index: { on_disk: false } } } }
        : {}),
    });

    return summaryFromAdvice(name, advice, sparseVectors, "created");
  }

  async ensure(name: string, options: AutoCreateOptions): Promise<CollectionSummary> {
    const collections = await this.client.getCollections();
    const exists = collections.collections.some((c) => c.name === name);

    if (exists) {
      const info = await this.client.getCollection(name);
      return {
        action: "already_exists",
        collection_name: name,
        vector_size: 0,
        distance: "",
        quantization: "",
        hnsw_m: undefined,
        hnsw_ef_construct: undefined,
        on_disk_payload: false,
        sparse_vectors: false,
        rationale: {},
        points_count: info.points_count ?? 0,
        status: info.status,
      };
    }

    return this.autoCreate(name, options);
  }

  async delete(name: string): Promise<boolean> {
    await this.client.deleteCollection(name);
    return true;
  }

  // ------------------------------------------------------------------ //
  // Payload indexes                                                     //
  // ------------------------------------------------------------------ //

  async createPayloadIndex(
    name: string,
    fieldName: string,
    fieldType: string = "keyword",
  ): Promise<PayloadIndexResult> {
    const validated = assertFieldType(fieldType);
    await this.client.createPayloadIndex(name, {
      field_name: fieldName,
      field_schema: validated,
    });
    return { collection: name, field: fieldName, type: fieldType, status: "indexed" };
  }

  // ------------------------------------------------------------------ //
  // Health and stats                                                    //
  // ------------------------------------------------------------------ //

  async health(name: string): Promise<HealthInfo> {
    const info = await this.client.getCollection(name);

    const result: HealthInfo = {
      collection_name: name,
      points_count: info.points_count ?? 0,
      indexed_vectors_count: info.indexed_vectors_count ?? 0,
      status: info.status,
      optimizer_status:
        typeof info.optimizer_status === "object" &&
        info.optimizer_status !== null &&
        "status" in info.optimizer_status
          ? String(info.optimizer_status.status)
          : String(info.optimizer_status),
    };

    // Decode VectorParams | Record<string, VectorParams>
    const vectors = info.config.params.vectors;
    if (vectors !== null && vectors !== undefined && typeof vectors === "object") {
      if ("size" in vectors && typeof vectors.size === "number") {
        result.vector_size = vectors.size;
        result.distance = String(vectors.distance);
      } else {
        // Named-vectors map — take first entry
        const first = Object.values(vectors)[0];
        if (
          first !== undefined &&
          typeof first === "object" &&
          first !== null &&
          "size" in first &&
          typeof first.size === "number"
        ) {
          result.vector_size = first.size;
          result.distance = String(first.distance);
        }
      }
    }

    const count = info.points_count ?? 0;
    if (!info.config.quantization_config && count > 100_000) {
      result.recommendation =
        `Collection has ${count.toLocaleString()} vectors but no quantization configured. ` +
        "Consider scalar INT8 quantization for ~4× memory reduction with minimal recall impact.";
    }

    return result;
  }

  // ------------------------------------------------------------------ //
  // Optimisation                                                        //
  // ------------------------------------------------------------------ //

  async optimize(name: string, priority: PriorityHint = "recall"): Promise<OptimizeResult> {
    const info = await this.client.getCollection(name);
    const count = info.points_count ?? 100_000;
    const optimizers = pickOptimizers(count, priority);

    await this.client.updateCollection(name, {
      optimizers_config: {
        indexing_threshold: optimizers.indexing_threshold,
        memmap_threshold: optimizers.memmap_threshold,
      },
    });

    return {
      collection: name,
      priority,
      applied_optimizers: {
        indexing_threshold: optimizers.indexing_threshold ?? 20_000,
        memmap_threshold: optimizers.memmap_threshold ?? 100_000,
      },
    };
  }
}

// ---------------------------------------------------------------------------
// Internal helpers
// ---------------------------------------------------------------------------

function summaryFromAdvice(
  name: string,
  advice: CollectionParamAdvice,
  sparseVectors: boolean,
  action: string,
): CollectionSummary {
  return {
    collection_name: name,
    vector_size: advice.vector_size,
    distance: advice.distance,
    quantization: advice.quantization_mode,
    hnsw_m: advice.hnsw_config.m,
    hnsw_ef_construct: advice.hnsw_config.ef_construct,
    on_disk_payload: advice.on_disk_payload,
    sparse_vectors: sparseVectors,
    rationale: advice.rationale,
    action,
  };
}
